sudo/lib/system/term/ops.rs

```diff
@@ -0,0 +1,186 @@
+    };
+    // Reset `GOT_SIGTTOU`.
+    GOT_SIGTTOU.store(false, Ordering::SeqCst);
+    // Set `action` as the action for `SIGTTOU` and store the original action in `original_action`
+    // to restore it later.
+    unsafe { sigaction(SIGTTOU, &action, original_action.as_mut_ptr()) };
+    // Call `tcsetattr` until it suceeds and ignore interruptions if we did not receive `SIGTTOU`.
+    loop {
+        match cerr(unsafe { tcsetattr(fd, flags, tp) }) {
+            Ok(_) => break,
+            Err(err) => {
+                let got_sigttou = GOT_SIGTTOU.load(Ordering::SeqCst);
+                if got_sigttou || err.kind() != io::ErrorKind::Interrupted {
+                    return Err(err);
+                }
+            }
+        }
+    }
+    // Restore the original action.
+    unsafe { sigaction(SIGTTOU, original_action.as_ptr(), std::ptr::null_mut()) };
+
+    Ok(())
+}
+
+/// Copy the settings of the `src` terminal to the `dst` terminal.
+pub fn term_copy<S: AsRawFd, D: AsRawFd>(src: &S, dst: &D) -> io::Result<()> {
+    let src = src.as_raw_fd();
+    let dst = dst.as_raw_fd();
+
+    let mut tt_src = MaybeUninit::<termios>::uninit();
+    let mut tt_dst = MaybeUninit::<termios>::uninit();
+    let mut wsize = MaybeUninit::<winsize>::uninit();
+
+    cerr(unsafe { tcgetattr(src, tt_src.as_mut_ptr()) })?;
+    cerr(unsafe { tcgetattr(dst, tt_dst.as_mut_ptr()) })?;
+
+    let tt_src = unsafe { tt_src.assume_init() };
+    let mut tt_dst = unsafe { tt_dst.assume_init() };
+
+    // Clear select input, output, control and local flags.
+    tt_dst.c_iflag &= !INPUT_FLAGS;
+    tt_dst.c_oflag &= !OUTPUT_FLAGS;
+    tt_dst.c_cflag &= !CONTROL_FLAGS;
+    tt_dst.c_lflag &= !LOCAL_FLAGS;
+
+    // Copy select input, output, control and local flags.
+    tt_dst.c_iflag |= tt_src.c_iflag & INPUT_FLAGS;
+    tt_dst.c_oflag |= tt_src.c_oflag & OUTPUT_FLAGS;
+    tt_dst.c_cflag |= tt_src.c_cflag & CONTROL_FLAGS;
+    tt_dst.c_lflag |= tt_src.c_lflag & LOCAL_FLAGS;
+
+    // Copy special chars from src verbatim.
+    tt_dst.c_cc.copy_from_slice(&tt_src.c_cc);
+
+    // Copy speed from src (zero output speed closes the connection).
+    {
+        let mut speed = unsafe { cfgetospeed(&tt_src) };
+        if speed == libc::B0 {
+            speed = libc::B38400;
+            unsafe { cfsetospeed(&mut tt_dst, speed) };
+            speed = unsafe { cfgetispeed(&tt_src) };
+            unsafe { cfsetispeed(&mut tt_dst, speed) };
+        }
+    }
+
+    tcsetattr_nobg(dst, TCSAFLUSH, &tt_dst)?;
+
+    cerr(unsafe { ioctl(src, TIOCGWINSZ, &mut wsize) })?;
+    cerr(unsafe { ioctl(dst, TIOCSWINSZ, &wsize) })?;
+
+    Ok(())
+}
+
+/// Set the `fd` terminal to raw mode. Enable terminal signals if `with_signals` is set to `true`.  
+pub fn term_raw<F: AsRawFd>(fd: &F, with_signals: bool) -> io::Result<()> {
+    let fd = fd.as_raw_fd();
+
+    if !unsafe { CHANGED } {
+        cerr(unsafe { tcgetattr(fd, OTERM.as_mut_ptr()) })?;
+    }
```

Comment:
if CHANGED is already true, should the code that follows (e.g. `tcsetattr_nobg`) still be called? IOW, what should happen if `term_raw` is executed twice one after the other? should the second call be a no-op or should it panic (i.e. that's a programmer error)

Reply:
The code that follow should still be called, or at least that's what ogsudo does. I think this is because ogsudo has other operations where `OTERM` is overwritten and then it does the call to `tcgetattr` to update the `termios` for `fd` and guarantee we always have the right `termios`.

I believe (with capital B) that we could simplify this code if we don't introduce other operation that plays with `CHANGED` and `OTERM`. In #363, I only needed what's implemented here.